### PR TITLE
allow cKDTree.query taking a list input in k.

### PR DESCRIPTION
**Pull request by rainwoodman**
the list input k will return only the neighbours ranks as
given in the list.

### CODE REVIEW

**Comment by ev-br** [comment]
What's the status of this after https://github.com/scipy/scipy/pull/4890 is merged?

**Comment by rainwoodman** [comment]
#4890 does not contain this.

The argument for the functionality (querying k-th-nn rather than all of k-nn) still applies. 

How do we like the proposed API changes?

- k can be an array-like for the list of requested k-th neighbours 
- integer input k is effectively converted to arange(k)
- when len(k) == 1, return value is squeezed. This preserves the undocumented current behavior of squeezing the result when k == 1.

**Comment by sturlamolden** [comment]
I am all for this PR. The current behavior is very wasteful with respect to search time and memory. Usually I only want to query for a single value of ```k``` or a small set of ```k```. I often do a search with ```k=12```, and this PR will be an order of magnitude improvement in terms of speed and memory if I can search with ```k=[12,]``` instead. 

I also like the idea of squeezing when len(k) == 1, because that is what is what I would need to do to post-process anyway.

**Comment by sturlamolden** [comment]
Ignore the comment I deleted. I was stupid. This PR can save memory for large data sets, but it will not save query time because k determines the size of the priority queue during the query. If we search for ```k=N``` we get the query for ```range(1,n)``` for free (except in terms of memory). I am still in favor of this because it does not do any harm, and will save memory when querying large data sets. But it is not a speed improvement over the current behavior (nor is it harmful). 

Sturla

**Comment by ev-br** [comment]
@rainwoodman @sturlamolden I'd be happy to hit the green button when you guys agree on this one. At a minimum, it needs a rebase.

**Comment by sturlamolden** [comment]
I am in favour of this PR. Usually I slice the output from ```.query``` anyway, so it will only make things easier. But @rainwoodman will have to do the rebase.

**Comment by sturlamolden** [comment]
I does not increase query speed though, but it has favorable memory characteristics for most usecases I am aware of.

**Comment by rainwoodman** [comment]
Indeed. I will rebase this this tomorrow!

File: scipy/spatial/ckdtree.pyx [hunk 1]
```diff
@@ -1175,8 +1175,9 @@ cdef class cKDTree:
         ----------
         x : array_like, last dimension self.m
             An array of points to query.
-        k : integer
-            The number of nearest neighbors to return.
+        k : integer or list
+            The number of nearest neighbors to return,
+            or a list of ranks of nearest neighbor to return.
```
Comment (rgommers) [inline scipy/spatial/ckdtree.pyx]: Adding an example of what this does and/or some more detailed explanation of the benefits in the ``Notes`` section would be good.

**Comment by rgommers** [comment]
> How do we like the proposed API changes?
>
>    - k can be an array-like for the list of requested k-th neighbours
>    - integer input k is effectively converted to arange(k)
>    - when len(k) == 1, return value is squeezed. This preserves the undocumented
      current behavior of squeezing the result when k == 1.

This looks fine to me.

**Comment by sturlamolden** [comment]
> - integer input k is effectively converted to arange(k)

Note that this is the current behaviour (actually ```arange(1,k)```).

With the proposed change we do a query for ```arange(1,max(k))``` and only store the columns for the requested values of k.

**Comment by rainwoodman** [comment]
I am working on a new patch and will push again. Not sure why github has automatically closed this..

**Comment by rgommers** [comment]
Allright, all green so in it goes. Thanks @rainwoodman!

**Comment by rainwoodman** [comment]
Thanks for merging!

**Pull request merged**